src/libmongoc/src/mongoc/mcd-azure.c

```diff
@@ -0,0 +1,167 @@
+   "&resource=https%3A%2F%2Fvault.azure.net";
+
+void
+mcd_azure_imds_request_init (mcd_azure_imds_request *req)
+{
+   _mongoc_http_request_init (&req->req);
+   // The HTTP host of the IMDS server
+   req->req.host = "169.254.169.254";
+   // No body
+   req->req.body = "";
+   // We GET
+   req->req.method = "GET";
+   // 'Metadata: true' is required
+   req->req.extra_headers = "Metadata: true\r\n"
+                            "Accept: application/json\r\n";
+   // The default path is suitable. In the future, we may want to add query
+   // parameters to disambiguate a managed identity.
+   req->req.path = bson_strdup (DEFAULT_METADATA_PATH);
+}
+
+void
+mcd_azure_imds_request_destroy (mcd_azure_imds_request *req)
+{
+   bson_free ((void *) req->req.path);
+   *req = (mcd_azure_imds_request){0};
+}
+
+bool
+mcd_azure_access_token_try_init_from_json_str (mcd_azure_access_token *out,
+                                               const char *json,
+                                               int len,
+                                               bson_error_t *error)
+{
+   bool okay = false;
+
+   if (len < 0) {
+      // Detect from a null-terminated string
+      len = strlen (json);
+   }
+
+   // Zero the output
+   *out = (mcd_azure_access_token){0};
+
+   // Parse the JSON data
+   bson_t bson;
+   if (!bson_init_from_json (&bson, json, len, error)) {
+      return false;
+   }
+
+   bson_iter_t iter;
+   // access_token
+   bool found = bson_iter_init_find (&iter, &bson, "access_token");
+   const char *const access_token =
+      !found ? NULL : bson_iter_utf8 (&iter, NULL);
+   // resource
+   found = bson_iter_init_find (&iter, &bson, "resource");
+   const char *const resource = !found ? NULL : bson_iter_utf8 (&iter, NULL);
+   // token_type
+   found = bson_iter_init_find (&iter, &bson, "token_type");
+   const char *const token_type = !found ? NULL : bson_iter_utf8 (&iter, NULL);
+
+   if (!(access_token && resource && token_type)) {
+      bson_set_error (
+         error,
+         MONGOC_ERROR_PROTOCOL_ERROR,
+         64,
+         "One or more required JSON properties are missing/invalid: data: %.*s",
+         len,
+         json);
+   } else {
+      // Set the output, duplicate each string
+      *out = (mcd_azure_access_token){
+         .access_token = bson_strdup (access_token),
+         .resource = bson_strdup (resource),
+         .token_type = bson_strdup (token_type),
+      };
+      okay = true;
+   }
+
+   bson_destroy (&bson);
+   return okay;
+}
+
+
+void
+mcd_azure_access_token_destroy (mcd_azure_access_token *c)
+{
+   bson_free (c->access_token);
+   bson_free (c->resource);
+   bson_free (c->token_type);
+}
+
+bool
+mcd_azure_send_request_with_retries (const mongoc_http_request_t *req,
+                                     mongoc_http_response_t *resp,
+                                     enum mcd_azure_http_flags flags,
+                                     bson_error_t *error)
```

Comment:
The retry logic is unfortunately ambiguously scoped. One _could_ say "send the request, retrying on these conditions, use exponential +2s backoff <=30s", and just assume that the correctness of that logic is outside of the scope of the document (and thus should be tested separately without being "part of" the Azure KMS logic).

We can't even be sure how common the retry would be needed. I only added it at the recommendation of the API docs.

It could also be specified as a "hint" recommendation, i.e. "drivers MAY add additional retry logic for the HTTP request."

Reply:
> The retry logic is unfortunately ambiguously scoped. One _could_ say "send the request, retrying on these conditions, use exponential +2s backoff <=30s", and just assume that the correctness of that logic is outside of the scope of the document (and thus should be tested separately without being "part of" the Azure KMS logic).

IMO specification changes should include guidance for testing. Otherwise it adds significant work to downstream teams to determine how to test.

> We can't even be sure how common the retry would be needed. I only added it at the recommendation of the API docs.

Good point. If there are no specification or prose tests, I say remove the retry logic. I expect adding any specification tests may be significant work (e.g. add a mock server). And at this point we do not know if there is any benefit.

> It could also be specified as a "hint" recommendation, i.e. "drivers MAY add additional retry logic for the HTTP request."

I think the general preference in the specifications is consistency between driver implementations (e.g. the [Strive to be idiomatic, but favor consistency](https://github.com/mongodb/specifications/tree/51741620a6af40337b8ec9dcb6fe44f8586f07f6#strive-to-be-idiomatic-but-favor-consistency) and [When in doubt, use "MUST" not "SHOULD](https://github.com/mongodb/specifications/tree/51741620a6af40337b8ec9dcb6fe44f8586f07f6#when-in-doubt-use-must-not-should-in-specs) mantras.

Rather than making it optional, I suggest pick one way (either with retry logic and tests, or no retry logic) and document the Rationale.